crates/loader/src/oci/cache.rs

```diff
@@ -0,0 +1,166 @@
+//! Cache for OCI registry entities.
+
+use anyhow::{bail, Context, Result};
+use oci_distribution::Reference;
+use tokio::fs;
+
+use std::path::{Path, PathBuf};
+
+const CONFIG_DIR: &str = "spin";
+const REGISTRY_CACHE_DIR: &str = "registry";
+const MANIFESTS_DIR: &str = "manifests";
+const WASM_DIR: &str = "wasm";
+const DATA_DIR: &str = "data";
+
+const MANIFEST_FILE: &str = "manifest.json";
+const CONFIG_FILE: &str = "config.json";
+const LATEST_TAG: &str = "latest";
+
+/// Cache for registry entities.
+pub struct Cache {
+    /// Root directory for the cache instance.
+    pub root: PathBuf,
+}
+
+impl Cache {
+    /// Create a new cache given an optional root directory.
+    pub async fn new(root: Option<PathBuf>) -> Result<Self> {
+        let root = match root {
+            Some(root) => root,
+            None => dirs::cache_dir()
+                .context("cannot get cache directory")?
+                .join(CONFIG_DIR),
+        };
+        let root = root.join(REGISTRY_CACHE_DIR);
+        Self::ensure_dirs(&root).await?;
+
+        Ok(Self { root })
+    }
+
+    /// The manifests directory for the current cache.
+    fn manifests_dir(&self) -> PathBuf {
+        self.root.join(MANIFESTS_DIR)
+    }
+
+    /// The Wasm bytes directory for the current cache.
+    fn wasm_dir(&self) -> PathBuf {
+        self.root.join(WASM_DIR)
+    }
+
+    /// The data directory for the current cache.
+    fn data_dir(&self) -> PathBuf {
+        self.root.join(DATA_DIR)
+    }
+
+    /// Return the path to a wasm file given its digest.
+    pub fn wasm_file(&self, digest: impl AsRef<str>) -> Result<PathBuf> {
+        let path = &self.wasm_dir().join(digest.as_ref());
+        match path.exists() {
+            true => Ok(path.into()),
+            false => bail!(format!(
+                "cannot find wasm file for digest {}",
+                digest.as_ref()
+            )),
+        }
+    }
+
+    /// Return the path to a data file given its digest.
+    pub fn data_file(&self, digest: impl AsRef<str>) -> Result<PathBuf> {
+        let path = &self.data_dir().join(digest.as_ref());
+        match path.exists() {
+            true => Ok(path.into()),
+            false => bail!(format!(
+                "cannot find data file for digest {}",
+                digest.as_ref()
+            )),
+        }
+    }
+
+    /// Get the file path to an OCI manifest given a reference.
+    /// If the directory for the manifest does not exist, this will create it.
+    pub async fn manifest_for_reference(&self, reference: impl AsRef<str>) -> Result<PathBuf> {
+        let reference: Reference = reference
+            .as_ref()
+            .parse()
+            .context("cannot parse reference")?;
```

Comment:
For this and the "cannot find" context (and same for the config function), suggest qualifying it as "OCI reference."  It is obvious at this point in the code but I don't know if it will be once it bubbles up to the command line.

Also, should they say which reference they are having problems with?  I suppose there should be only one OCI reference in play...?